File: 3-longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
    cmap := make(map[byte]int)
    max_length, start := 0, -1
    for idx := 0; idx != len(s); idx += 1{
        ch := s[idx]
        last_idx, ok := cmap[ch]
        if ok && last_idx > start{
            start = last_idx
        }
        cmap[ch] = idx
        if idx - start > max_length{
            max_length = idx - start
        }
    }
    return max_length
}

File: captainhcg/js/3-longest-substring-without-repeating-characters.js
var lengthOfLongestSubstring = function(s) {
    var lastMap = new Map();
    var maxLen = 0;
    var lastIdx = -1;
    for(var idx=0; idx < s.length; idx++){
        var ch = s[idx];
        if(lastMap.has(ch)){
            lastIdx = Math.max(lastIdx, lastMap.get(ch));
        }
        lastMap.set(ch, idx);
        maxLen = Math.max(maxLen, idx - lastIdx);
    }
    return maxLen;
};
